crates/shared/src/generator.rs

```diff
@@ -0,0 +1,164 @@
+use futures::{channel::mpsc, stream::FusedStream, SinkExt, Stream};
+use std::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+/// Creates a stream from an async generator function.
+///
+/// This is similar to [`unfold`](https://docs.rs/futures/0.3.21/futures/stream/fn.unfold.html) in
+/// the sense that it bridges Future to Stream. In addition, it can be more ergonomic because it
+/// allows the compiler to automatically create the state machine that you would otherwise have to
+/// write by hand in the same way that async/await syntax makes it easier to create a Future.
+///
+/// The generator takes an argument `Sender` who's `send` function is used to yield values to the
+/// Stream.
+///
+/// The generator must ensure that the sender does not outlive the generator. This is usually the
+/// case but could be violated by leaking the sender or moving it into a global task. If you want to
+/// use a task then you do not need this function because you can communicate over a channel
+/// directly.
+pub fn async_generator_to_stream<T, Fut>(
+    generator: impl FnOnce(Sender<T>) -> Fut,
+) -> impl Stream<Item = T> + FusedStream
+where
+    Fut: Future<Output = ()>,
+{
+    let (sender, receiver) = mpsc::channel(0);
```

Comment:
If we were to have this buffered, that just means that it is possible for the `Future` to produce more values than actually get consumed by the `Stream` right?